### DISCUSSION TITLE
Testing instrumentation with the reltest

### DISCUSSION THREAD

**Discussion by Rudis1261** [go-rel/rel#376] [Q&A] [answered]
I am still fairly new to golang especially the testing side of things. I have been digging through the source code and I wanted to see how I can test an instrumentation middleware for echo I have. 

To me it seems like it gets stubbed out in the reltest repository mock.

For reference this is the middleware I am talking about, it adds queries for debugging for admins in my application:
```golang
func RelQueryLoggerMiddleware(next echo.HandlerFunc) echo.HandlerFunc {
	return func(c echo.Context) error {
		// Don't do any of this when public assets
		if ignorePublicAssetRoutes(c) {
			return next(c)
		}

		templateData := c.Get("tmpl").(map[string]interface{})
		debugging, ok := c.Get("sql").(db.Debugging)
		if !ok {
			debugging = db.Debugging{}
		}

		// Log the executed SQL queries
		db.Database.Instrumentation(func(ctx context.Context, op string, message string, args ...interface{}) func(err error) {
			t := time.Now()

			return func(err error) {
				if op != "adapter-query" {
					return
				}

				duration := time.Since(t)
				debugging.Queries = append(debugging.Queries, db.Query{
					Statement: message,
					Duration:  duration,
					Err:       err,
					Args:      args,
				})

				if err != nil {
					debugging.Errors += 1
				}

				debugging.Duration = debugging.Duration + duration
				helper.SetDBDebugging(templateData, debugging)
			}
		})

		return next(c)
	}
}
```

Is this even a sane approach?

**Comment by Rudis1261** [comment]
![Screenshot 2024-10-23 at 07 38 12](https://github.com/user-attachments/assets/7b1164fb-4106-4010-8dd8-67e86af8a673)

**Comment by Fs02** [comment, marked as answer]
you can extract the function inside instrumentation as standalone function, and test it separately

**Reply by Rudis1261** [reply]
You are probably correct. Store them in the context perhaps, something like that 🤔 

Would you be able to point out in the source code where this instrumentation adds the queries so I can better understand when it fires, I tried to find it but must have missed it?

**Reply by Fs02** [reply]
for example: https://github.com/search?q=org%3Ago-rel+Observe&type=code

**Reply by Rudis1261** [reply]
Thanks so much! And thanks for the awesome lib again! Love the reltest package it makes testing a breeze, and in general I love working with rel!
